nbs/01_io.ipynb

```diff
@@ -266,6 +268,148 @@
+    "    create = dict_with_create['create']\n",
+    "    name = create.get('name', f\"{dict_with_create['name']}_{create['type']}\")\n",
+    "    # Compile regex to catch df.columns that we want to apply topk for.\n",
+    "    regex_from = re.compile(create['from_columns'])\n",
+    "    from_cols = list(filter(lambda s: regex_from.match(s), df.columns))\n",
+    "    # agg_ind is index of the regex group that we want to aggregate over.\n",
+    "    # All other indeces are unique identifiers for each subgroup\n",
+    "    # e.g. ['A', 'a'] and ['B', 'b'] (that are also added to column names)\n",
+    "    # Recall that regex group 0 is the whole match.\n",
+    "    # Note that re.Match.groups() does not include the whole match.\n",
+    "    # This means that we need to subtract 1 from agg_ind.\n",
+    "    agg_ind = create.get('agg_index', -1)\n",
+    "    agg_ind = agg_ind-1 if agg_ind > 0 else agg_ind\n",
+    "    n_groups = len(regex_from.match(from_cols[0]).groups())\n",
+    "    has_subgroups = n_groups >= 2 # Multiple aggregations needed?\n",
+    "    regex_to = create.get('res_cols', '')\n",
+    "    kmax = create.get('k',None)\n",
+    "    na_vals = create.get('na_vals', [])\n",
+    "    if has_subgroups:\n",
+    "        # collect all subgroups, later aggregate each subgroup separately\n",
+    "        def get_subgroup_id(column):\n",
+    "            \"Discard the aggregation index and collect the rest as identifier.\"\n",
+    "            subgroup_id = list(regex_from.match(column).groups())\n",
+    "            subgroup_id.pop(agg_ind)\n",
+    "            return subgroup_id\n",
+    "        subgroups_ids = np.unique(\n",
+    "            np.vstack(list(map(get_subgroup_id,from_cols))),\n",
+    "            axis=0\n",
+    "            )\n",
+    "        subgroups = [\n",
+    "            [col for col in from_cols if get_subgroup_id(col)==subgroup_id]\n",
+    "            for subgroup_id in subgroups_ids\n",
+    "            ]\n",
+    "    else: subgroups = [from_cols]\n",
+    "    topk_dfs, subgroup_metas = [], []\n",
+    "    # replace column name with regex group to allow translate later in io.py\n",
+    "    get_regex_group_at_agg_ind = lambda s: regex_from.match(s).groups()[agg_ind]\n",
+    "    for subgroup in subgroups:\n",
+    "        sdf = df[subgroup].astype('object').replace(na_vals,None)\n",
+    "        newcols = [ # new col names mapped with regex groups from orig col names\n",
+    "            regex_from.match(col).expand(regex_to) for col in sdf.columns\n",
+    "            ]\n",
+    "        # next, replace col names with only the regex group at agg_ind\n",
+    "        # fun is defined inside a bigger fun, thus agg_ind acts as a global var\n",
+    "        sdf.columns = sdf.columns.map(get_regex_group_at_agg_ind)\n",
+    "        sdf = sdf.where(sdf.isna(), other=pd.Series(sdf.columns, index=sdf.columns), axis=1) # replace cell with column name where not NA\n",
```

Comment:
Nit: I'm used to .mask instead of .where (actually had to look up what the latter does). 
I prefer mask because it tells which rows it changes, which feels more intuitive than giving the ones you leave the same....